### PR TITLE
Master orgchart fp tde

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
Task: https://www.odoo.com/web#id=32086&view_type=form&model=project.task&action=333&active_id=248&menu_id=4720

### CODE REVIEW

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    <t t-if="widget.emp_data.children.length || widget.emp_data.managers.length" t-call="hr_org_chart_employee">
+        <t t-set="employee_type" t-value="'self'"/>
+        <t t-set="employee" t-value="widget.emp_data.self"/>
+    </t>
+
+    <t t-if="!widget.emp_data.children.length &amp;&amp; !widget.emp_data.managers.length">
+        <div class="alert alert-info">
+            <p><b>No hierarchy position.</b></p>
+            <p>This employee has not managers or subordinates defined.</p>
```
Comment (bst-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: This employee has no managers or subordinates.

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            <p>In order get the organigram, assign a manager or set the employee as a manager for others.</p>
```
Comment (bst-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: In order to get an organigram, assign a manager to the related employees.

File: addons/hr_org_chart/controllers/hr_org_chart.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            indirect_sub_count=employee.child_all_count,
+        )
+
+    @http.route('/hr/get_org_chart', type='json', auth='user')
+    def get_org_chart(self, employee_id):
+        if not employee_id:  # to check
+            return {}
+
+        employee = request.env['hr.employee'].browse(employee_id)
+        # check and raise
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/controllers/hr_org_chart.py]: I forgot to add some security and redirection. I will add them.

File: addons/hr_org_chart/views/hr_templates.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+        <template id="assets_backend" name="tours backend assets" inherit_id="web.assets_backend">
+            <xpath expr="//script[last()]" position="after">
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/views/hr_templates.xml]: Don't know, original fp code. I will update it !

File: addons/hr_org_chart/static/src/js/hr_org_chart.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    render_value: function () {
+        if (! this.view.datarecord.id) {
+            return this.$el.html(QWeb.render("hr_org_chart", {widget: this}));
+        }
+
+        var self = this;
+        this.get_org_chart_data(this.view.datarecord.id).then(function () {
+            self.$el.html(QWeb.render("hr_org_chart", {widget: self}));
+        }).then(function () {
+            self.$el.find('[data-toggle="popover"]').each(function () {
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/js/hr_org_chart.js]: "some" time. Grmglblgl.

File: addons/hr_org_chart/static/src/js/hr_org_chart.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    on_employee_sub_redirect: function (event) {
+        event.preventDefault();
+        var employee_id = parseInt($(event.currentTarget).data('employee-id'));
+        var employee_name = $(event.currentTarget).data('employee-name');
+        var type = $(event.currentTarget).data('type') || 'direct';
+        var domain = [['parent_id', '=', employee_id]];
+        var name = _t('Direct Subordinates of ' + employee_name);
+        if (type === 'total') {
+            domain = ['&', ['parent_id', 'child_of', employee_id], ['id', '!=', employee_id]];
+            name = _t('Subordinates of ' + employee_name);
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/js/hr_org_chart.js]: Nice.

File: addons/hr_org_chart/views/hr_views.xml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+<odoo>
+    <record id="hr_employee_view_form_inherit_org_chart" model="ir.ui.view">
+        <field name="name">hr.employee.view.form.inherit.org_chart</field>
+        <field name="model">hr.employee</field>
+        <field name="inherit_id" ref="hr.view_employee_form"/>
+        <field name="arch" type="xml">
+            <div id="o_employee_main" position="after">
+                <div id="o_employee_right">
+                    <h4 class="o_org_chart_title mb16 mt0">Organization Chart</h4>
+                    <field name="child_ids" widget="hr_org_chart" options="{'managers_level': 2}"/>
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/views/hr_views.xml]: Either we put it at 17, either we remove it. Let us remove it. Good catch.

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml (hunk 1 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: Done !

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml (hunk 2 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: Done !

File: addons/hr_org_chart/views/hr_templates.xml (hunk 4 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/views/hr_templates.xml]: Done !

File: addons/hr_org_chart/static/src/js/hr_org_chart.js (hunk 6 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/js/hr_org_chart.js]: Done !

File: addons/hr_org_chart/views/hr_views.xml (hunk 7 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/views/hr_views.xml]: Done !

File: addons/hr_org_chart/controllers/hr_org_chart.py (hunk 3 above)
Comment (tde-banana-odoo) [inline addons/hr_org_chart/controllers/hr_org_chart.py]: Done !

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                        Total</a>
+                </td>
+            </tr>
+        </tbody>
+    </table>
+</t>
+
+<t t-name="hr_orgchart_emp_popover_title">
+    <div>
+        <span t-att-style='"background-image:url(\"/web/image/hr.employee/" + employee.id + "/image/\")"'/>
```
Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: Mpf. I trust my designer :) .

File: addons/hr_org_chart/static/src/xml/hr_org_chart.xml [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    <t t-if="widget.emp_data.children.length || widget.emp_data.managers.length" t-call="hr_org_chart_employee">
+        <t t-set="employee_type" t-value="'self'"/>
+        <t t-set="employee" t-value="widget.emp_data.self"/>
+    </t>
+
+    <t t-if="!widget.emp_data.children.length &amp;&amp; !widget.emp_data.managers.length">
+        <div class="alert alert-info">
+            <p><b>No hierarchy position.</b></p>
+            <p>This employee has no manager or subordinate.</p>
+            <p>In order get an organigram, assign a manager to the related employees.</p>
```
Comment (bst-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: In order **TO** get an organigram. 

@tde-banana-odoo A word is missing. Maybe YTI abducted it, it was quite young.

Comment (tde-banana-odoo) [inline addons/hr_org_chart/static/src/xml/hr_org_chart.xml]: Too late my dear. You will have to fill the form in order to make a fix in master.

**Pull request merged**